src/viem/prepareTransactions.ts

```diff
@@ -0,0 +1,314 @@
+import erc20 from 'src/abis/IERC20'
+
+// Apply a multiplier for the decreased swap amount to account for the
+// varying gas costs of different swap providers (or even the same swap)
+const DECREASED_SWAP_AMOUNT_GAS_COST_MULTIPLIER = 1.2
+
+interface PreparedTransactionsPossible {
+  type: 'possible'
+  transactions: TransactionRequestCIP42[]
+}
+
+export interface PreparedTransactionsNeedDecreaseSpendAmountForGas {
+  type: 'need-decrease-spend-amount-for-gas'
+  maxGasCost: BigNumber
+  feeCurrency: TokenBalance
+  decreasedSpendAmount: BigNumber
+}
+
+export interface PreparedTransactionsNotEnoughBalanceForGas {
+  type: 'not-enough-balance-for-gas'
+  feeCurrencies: TokenBalance[]
+}
+
+export type PreparedTransactionsResult =
+  | PreparedTransactionsPossible
+  | PreparedTransactionsNeedDecreaseSpendAmountForGas
+  | PreparedTransactionsNotEnoughBalanceForGas
+
+export function getMaxGasCost(txs: TransactionRequestCIP42[]): BigNumber {
+  let maxGasCost = BigInt(0)
+  for (const tx of txs) {
+    if (!tx.gas || !tx.maxFeePerGas) {
+      throw new Error('Missing gas or maxFeePerGas')
+    }
+    maxGasCost += BigInt(tx.gas) * BigInt(tx.maxFeePerGas)
+  }
+  return new BigNumber(maxGasCost.toString())
+}
+
+function getFeeCurrencyAddress(feeCurrency: TokenBalance) {
+  return !feeCurrency.isNative ? (feeCurrency.address as Address) : undefined
+}
+
+async function tryEstimateTransaction({
+  baseTransaction,
+  maxFeePerGas,
+  feeCurrencyAddress,
+  maxPriorityFeePerGas,
+}: {
+  baseTransaction: TransactionRequestCIP42
+  maxFeePerGas: bigint
+  feeCurrencyAddress?: Address
+  maxPriorityFeePerGas?: bigint
+}) {
+  const tx = {
+    ...baseTransaction,
+    maxFeePerGas,
+    maxPriorityFeePerGas,
+    // Don't include the feeCurrency field if not present.
+    // See https://github.com/wagmi-dev/viem/blob/e0149711da5894ac5f0719414b4ecc06ccaecb7b/src/chains/celo/serializers.ts#L164-L168
+    ...(feeCurrencyAddress && { feeCurrency: feeCurrencyAddress }),
+  }
+
+  // TODO maybe cache this? and add static padding when using non-native fee currency
+  try {
+    tx.gas = await publicClient.celo.estimateGas({
+      ...tx,
+      account: tx.from,
+    })
+  } catch (e) {
+    if (e instanceof EstimateGasExecutionError) {
+      // Likely too much gas was needed
+      Logger.warn(
+        'SwapScreen@useSwapQuote',
+        `Couldn't estimate gas with feeCurrencyAddress ${feeCurrencyAddress}`,
+        e
+      )
+      return null
+    }
+    throw e
+  }
+
+  return tx
+}
+
+async function tryEstimateTransactions(
+  baseTransactions: TransactionRequestCIP42[],
+  feeCurrency: TokenBalance
+) {
+  const transactions: TransactionRequestCIP42[] = []
+
+  const feeCurrencyAddress = getFeeCurrencyAddress(feeCurrency)
+  const { maxFeePerGas, maxPriorityFeePerGas } = await estimateFeesPerGas(
+    publicClient.celo,
+    feeCurrencyAddress
+  )
+
+  for (const baseTx of baseTransactions) {
+    if (baseTx.gas) {
```

Comment:
this line was changed. previously we just estimated gas for the first tx, now makes gas estimation possible for any of the transaction inputs that don't already have a gas field

Reply:
This looks good in most cases.

I'm trying to anticipate incorrect usage of this function.
Let's say someone calls it with multiple dependant TXs (like for swaps) and doesn't set the `gas` for any of them.
The first estimate will work, and then for the next TX it should likely fail (because it needed the first TX to have been executed), but then will just try the next fee currency and fail again.
And eventually it will return a `not-enough-balance-for-gas` result.
Which isn't correct.

But now I'm thinking we could maybe improve how we handle `if (e instanceof EstimateGasExecutionError) {` and look at the exact error message instead and if it's really about not enough gas, we can indeed try the next fee currency, otherwise we must abort and throw that we can't estimate (i.e. the dependant TXs case without gas set)

I'm happy to try this in a follow up PR not to hold this one more than necessary.
What do you think?

Reply:
it does sound like in the ideal case, we would handle it more intelligently when gas estimation fails because the transaction failed for reasons unrelated to gas (in this case, because the transferFrom amount wasn't approved yet for the swap contract)

Reply:
if you're willing to give it a shot I'd appreciate it. I do think being able to handle independent transfers will be helpful for DEK registration